### PR TITLE
Implementation of Tuples

### PR DESCRIPTION
**Pull request by TAswan**
Majority of Tuple features implemented: 

1.  added AST and internal representations (Domains etc)
2. added parsing of tuple variables and literals
3. added representation rules for Tuples similarly to matrices
4. added rules for indexing, equality and inequality

Features not implemented:

1. nesting tuples - due to current issues with representation that will require a larger rework
2. comparitives on tuples (leq, geq, etc) - due to no current support for mkDotLeqOp etc. unsure how big a change this is, but rules can be added fairly easily if these operations are trivial to add.

### CODE REVIEW

File: crates/conjure_core/src/rules/tuple.rs [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+    let decl = symbols.lookup(name).unwrap();
+
+    let Some(Domain::DomainTuple(_)) = decl.domain().cloned().map(|x| x.resolve(symbols)) else {
+        return Err(RuleNotApplicable);
+    };
+
+    let mut indices_are_const = true;
+    let mut indices_as_lits: Vec<Literal> = vec![];
+
+    for index in indices {
```
Comment (niklasdewally) [inline crates/conjure_core/src/rules/tuple.rs:48]: Can tuples be multi-dimensional (which would mean they could be indexed by x[1,2]), or are they always one dimensional (x[1])?

If they are always one dimensional, this could be simplified further.

File: crates/conjure_core/src/rules/tuple.rs [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+            name.clone(),
+            "tuple_to_atom".into(),
+            indices_as_lits.iter().join("_"),
+        );
+
+        let subject = repr.expression_down(symbols)?[&indices_as_name].clone();
+
+        Ok(Reduction::pure(subject))
+    } else {
+        todo!("can tuples even do this?")
```
Comment (niklasdewally) [inline crates/conjure_core/src/rules/tuple.rs:67]: e.g. x[y], where x is a tuple and y is a decision variable

You can do the same thing as matrices here, and return a 1d matrix literal indexed by an index.

For example

```
find x: tuple(bool,bool)
find y: int(1..5)
such that
x[y]
---
x#tuple_to_atom[y] ~~~> [x#tuple_to_atom_1,x#tuple_to_atom_2][y]
```

You need to put it in the following form so that it can be translated to Minion
https://github.com/conjure-cp/conjure-oxide/blob/90d15b1c60db7b01dc87ddacd1efa4415f2bba69/crates/conjure_core/src/rules/minion.rs#L730-L735

Unlike matrices, I think you can assume that all indices you get as input 1-indexed, and that tuples cannot have multiple indices like x[1,2]: under these conditions, all you need to do is rewrite the subject to a matrix literal of its elements, and use the same index. (For matrices, we have to both flatten the list, and convert the indices to 1-indexing here)

**Review by niklasdewally — COMMENTED** [review]
On the whole LGTM, left some comments below:

* Could you add a test case showing what happens when x is indexed by something out of range? (to test the definedness semantics, and bubbling)

**Review by ozgurakgun — COMMENTED** [review]
hi @TAswan - this is looking good overall. can you address the comments @niklasdewally left and update the PR please and let me know please? I'll then take another pass.

File: crates/conjure_core/src/rules/tuple.rs (hunk 2 above)
Comment (TAswan) [reply crates/conjure_core/src/rules/tuple.rs:67]: This part has been removed rather than implementing your suggestion, as conjure doesn't support tuple indexing by decision variable due to tuples being able to return different types based on the value given (a poor man's dependent types). While I think implementing in this way is possible, it would probably lead to bigger issues in the future (especially if we plan on type checking).

**Comment by ozgurakgun** [comment]
> tested, but haven't added to the test suite, as in both cases it fails at the conjure parser. This may be something we want to check later once the tree sitter parser supports tuples.

can you give me an example input and command I can run that will show the parser error please?

**Comment by TAswan** [comment]
@ozgurakgun

Not at my laptop so can't double check but it was

```
Find x: tuple (bool, bool)
such that
x[3] = false
```

**Comment by TAswan** [comment]
On laptop now, can confirm it was the above to check out of bounds, and in both conjure and oxide it gives the following

```
  |
3 | x[3]
  |   ^
Error: Tuple index 3 out of bounds
```

**Comment by ozgurakgun** [comment]
Ooops, `conjure pretty` is doing more checking than I thought it would be doing. I guess testing the out of bounds indexing can wait until the tree-sitter parser is ready to parse tuples then.

**Review by ozgurakgun — APPROVED** [review]
